### ISSUE TITLE
GNOME 44: `Super + Shift + Above_Tab` works in the forward direction when it should work in the backward direction

### ISSUE THREAD

**Issue by letavocado** [paperwm/PaperWM#523]
**Describe the bug**
<kbd>Super</kbd> <kbd> Shift</kbd>  <kbd>Above_Tab</kbd> works in the forward direction when it should work in the backward direction. Thus making no difference between the behavior of <kbd>Super</kbd> <kbd>Above_Tab</kbd> and <kbd>Super</kbd> <kbd> Shift</kbd>  <kbd>Above_Tab</kbd>

While <kbd>Super</kbd> <kbd> Ctrl</kbd> <kbd> Shift</kbd>  <kbd>Above_Tab</kbd> works as expected in backward cycle direction. Now the behavior works correctly, but sometimes the bug occurs.

**To Reproduce**
1. Switch to [fixes-gnome-44-support](https://github.com/paperwm/PaperWM/tree/fixes-gnome-44-support)
2. run './install.sh'
3. do a login/logout (maybe twice, if bug not occurred)
4. check behavior of described bug 

**Expected behavior**
<kbd>Super</kbd> <kbd> Shift</kbd>  <kbd>Above_Tab</kbd> should work in the backward direction 

**System information:**
`./gather-system-info.sh` output:

Distribution: Arch Linux
GNOME Shell 44.1
Display server: Wayland
PaperWM branch/tag: fixes-gnome-44-support
PaperWM commit: 477d546e5a78280cb324379a365225b0f702ad8d
Enabled extensions:
- paperwm@hedning:matrix.org
- clipboard-history@alexsaveau.dev
- blur-my-shell@aunetx
- gnome-ui-tune@itstime.tech
- Vitals@CoreCoding.com
- trayIconsReloaded@selfmade.pl
- switcher@landau.fi
- dash-to-dock@micxgx.gmail.com

**Additional context**
<kbd>Super</kbd> <kbd> Ctrl</kbd> <kbd> Shift</kbd>  <kbd>Above_Tab</kbd> works as expected in backward cycle direction. Now the behavior works correctly, but sometimes the bug occurs.

**Comment by jtaala** [comment]
Thanks @letavocado, haven't been able to reproduce as yet - but will try again later on.

I noted you're on a (slightly) previous commit.  I made some changes to stack mode (fixed another issue) today - can you pull that latest commit  https://github.com/paperwm/PaperWM/pull/515/commits/7358da5466d0f03551c0036fd2da3f3bb4f7e37a and see if you have any luck reproducing this?

Cheers!

**Comment by jtaala** [comment]
(although I doubt those changes I made would fix/impact this issue... was only to stack mode when mouse clicking on a window).

**Comment by letavocado** [comment]
pulled and the bug is here again

---
Distribution: Arch Linux
GNOME Shell 44.1
Display server: Wayland
PaperWM branch/tag: fixes-gnome-44-support
PaperWM commit: 7358da5466d0f03551c0036fd2da3f3bb4f7e37a
Enabled extensions:
- paperwm@hedning:matrix.org
- clipboard-history@alexsaveau.dev
- blur-my-shell@aunetx
- gnome-ui-tune@itstime.tech
- Vitals@CoreCoding.com
- trayIconsReloaded@selfmade.pl
- switcher@landau.fi
- dash-to-dock@micxgx.gmail.com

**Comment by letavocado** [comment]
Hmmmm...

I did `login/logout again`, but bug still exist
I locked screen and unlocked. And bug disappeared (Ta-da! It's gone).

**Comment by jtaala** [comment]
Interesting, that lock/unlock part kind of reminds me of some behaviours that #477 fixed.

Okay, to help me understand this a bit better (I don't think I'm understanding it right atm - maybe because I don't really use that functionality... maybe I should).

If you're feeling artistic, can you do up a screen capture (even if it's on your phone) of what it should look like (e.g. when the bug doesn't occur) - just with a few windows or something.  Alternatively, if the bug is happening can screen capture that too.  Just so I can compare to what I'm seeing.

Thanks @letavocado!

**Comment by letavocado** [comment]
Since github has a 10 mb video size limit, so the video is of poor quality.

https://github.com/paperwm/PaperWM/assets/31704887/3ab01b9b-7f00-44bf-88e8-2df0ab7960a1

**Comment by letavocado** [comment]
I can record it again or we can call via google meets tomorrow.

**Comment by jtaala** [comment]
> I can record it again or we can call via google meets tomorrow.

All good - that's perfect.

It actually sounds like a keybind clash - we had another one in gnome 44 (the default tile left/right keybinds).  Might not be, but in any case your vid is perfect and def enough to go on.  Will check it out tonight.

Cheers.

**Comment by letavocado** [comment]
Thanks @jtaala!

**Comment by jtaala** [comment]
Hey @letavocado - I've been trying and no matter what I can't reproduce this.

My current hypothesis is that there's a keybind clash between `super+above_tab` in PaperWM and another extension you're using that causing some funny behaviour.

This makes a little sense since after lock/unlock, PaperWM reapplies keybinds (which in theory would overwrite other extension keybinds... although that depends on the order gnome loads the extensions).

For testing purposes, can you please either

- change PaperWM keybinds to something else for those operations and see if you can reproduce this;
- or disable other extensions and see if you can reproduce (then if you have time, you could enable other extensions one by one to try isolate which one it might be);

The first one is probably a bit easier.

Thanks!

**Comment by jtaala** [comment]
I'll keep trying to reproduce - if anyone else sees this and has this issue please let me know.

**Comment by letavocado** [comment]
Hey, @jtaala. Feel free to ask me anything.

I added this. <kbd>Super</kbd><kbd>1</kbd> is handy, but sometimes this keybinding starts working after the second press. So I also added <kbd>Super</kbd> <kbd>Backslash</kbd>, which works immediately. 

![image](https://github.com/paperwm/PaperWM/assets/31704887/073e8a1e-885a-407d-b27d-2ba4e9f73bef)

Adding new keybinding not fixed the bug. 
At first it seemed that now it works correctly. I did a logout/login and now, figured out the new bug. When switching workspaces in the forward direction via <kbd>ctrl</kbd><kbd>above_tab</kbd>, the active windows began to move, as if I was holding the shift, although I did not hold the shift. 
 
I also tried disabling extensions. I did a several logout/login. But, nothing changed, bug still exist.

Distribution: Arch Linux
GNOME Shell 44.1
Display server: Wayland
PaperWM branch/tag: fixes-gnome-44-support
PaperWM commit: 7358da5466d0f03551c0036fd2da3f3bb4f7e37a
Enabled extensions:
- paperwm@hedning:matrix.org
- switcher@landau.fi
- dash-to-dock@micxgx.gmail.com

**Comment by letavocado** [comment]
I'll check it out a bit later and report back. Maybe I'll do a complete reinstall of PaperWM(or should I not?).

**Comment by jtaala** [comment]
Thanks @letavocado - okay.  

I noticed you've still got dash-to-dock running.  Just for this test, please disable that and maybe do a restart just to be sure.

Also try remove/delete the other shortcuts (so you only have one) - e.g with the back-x button there.

It's very strange, I haven't been able to reproduce this at all after many attempts.  I've reached out to others too to see if anyone else is having this issue.

**Comment by jtaala** [comment]
Okay, this is strange - I installed dash-to-dock and after enabling it and using that shortcut I had some real weird behaviour and it started grabbing the windows instead.

Even after disabling dash-to-dock, behaviour was still there - had to lock/unlock to get back to normal.

I'm guessing there's some kind of weird clash going on with that shortcut and dash-to-dock.

**Comment by letavocado** [comment]
I pulled the latest commit. Disabled dash-to-dock and rebooted the system. The bug still appears. What's more, the default keybinding <kbd>Super</kbd><kbd>Shift</kbd><kbd>Above_Tab</kbd> still works even though I removed it, as you can see in the attached screenshot.

![image](https://github.com/paperwm/PaperWM/assets/31704887/2374f54f-2b76-40ea-89bd-9e4329100cf0)

---
Distribution: Arch Linux
GNOME Shell 44.1
Display server: Wayland
PaperWM branch/tag: fixes-gnome-44-support
PaperWM commit: 477e87a4f04a01ad05245852c0cf645fd19bec0c
Enabled extensions:
- paperwm@hedning:matrix.org
- switcher@landau.fi

**Comment by jtaala** [comment]
> still works even though I removed it, as you can see in the attached screenshot.

Yes, that is strange and shouldn't happen.

Okay, let's remove PaperWM and do a reinstall.  Which mean might need to git clone it again, switch to the `fixes-gnome-44-support` branch and reinstall.

**Comment by jtaala** [comment]
e.g.

```
git clone git@github.com:paperwm/PaperWM.git
cd PaperWM
git checkout fixes-gnome-44-support
./install.sh
```

**Comment by jtaala** [comment]
Well, at least we're getting close to reproducing the weird stuff - I'll work on reproducing what I saw again and see where that leads.

**Comment by letavocado** [comment]
![image](https://github.com/paperwm/PaperWM/assets/31704887/b35ddcaa-6823-4e3d-977e-001ad9dc345a)

```
Use recommended settings (generates a backup) [Y/n]: y
Changed org.gnome.mutter.keybindings toggle-tiled-left from '@as []' to '[]'
Changed org.gnome.mutter.keybindings toggle-tiled-right from '@as []' to '[]',
```

maybe this is the reason of the bug? (Hmmm... smth about keybindings :thinking: )

**Comment by jtaala** [comment]
Well, that's fixing a keybinding clash with gnome 44 (see #518) - that shouldn't be it though (it just disables the default toggle-tiled-left/right shortcuts in gnome 44.

With the reinstall, are you still getting that bug?

**Comment by jtaala** [comment]
nevermind, the weird issue I saw was because I had a clash (I also `super+shift+~` for moving window to previous workspace).

Let me know how you go with the reinstall.

**Comment by letavocado** [comment]
Yes, I'm still getting the bug. After reinstalling, my settings remained the same (workspace names, keybinding). How can I totally reinstall? Where can I find user configs?

---
Distribution: Arch Linux
GNOME Shell 44.1
Display server: Wayland
PaperWM branch/tag: fixes-gnome-44-support
PaperWM commit: 477e87a4f04a01ad05245852c0cf645fd19bec0c
Enabled extensions:
- paperwm@hedning:matrix.org
- switcher@landau.fi

**Comment by letavocado** [comment]
I even tried disabling switcher. Same bug.

**Comment by letavocado** [comment]
I locked the screen five times, the bug didn't go away like it usually did before.

**Comment by jtaala** [comment]
User config is in dconf.

You should be able to delete them with this command:

```
dconf reset -f "/org/gnome/shell/extensions/paperwm/"
```

**Comment by jtaala** [comment]
Is the old shortcut still working (even though you've changed it)?

I'm back at the stage where I can't reproduce any of this.  We'll have to reach out to others to see if they're seeing this behaviour.

On a sidenote, is the shortcuts `Super+PageUp/PageDown` showing this behaviour (this is the shortcuts for switching to workspace above/below).

**Comment by letavocado** [comment]
Totally reinstalled, also removed `./.config/paperwm/`

and used this command
```
dconf reset -f "/org/gnome/shell/extensions/paperwm/"
```

settings was reset(except workspace names)

![image](https://github.com/paperwm/PaperWM/assets/31704887/95a22892-535f-465c-b21f-454202966011)

> On a sidenote, is the shortcuts Super+PageUp/PageDown showing this behaviour (this is the shortcuts for switching to workspace above/below).

In my case I additionally held down <kbd>fn</kbd> and pressed <kbd>up</kbd> or <kbd>down</kbd>

![image](https://github.com/paperwm/PaperWM/assets/31704887/30067df7-2918-474f-8053-5df0df2474c8)

Unfortunately, there was a bug here too. The `Super+PageUp` does not work at all, only the `Super+PageDown`. And when it reaches the last workspace at all stops working both of them(`PageUp`and `PageDown`). In workspaces above latest, also works only `PageDown`.


Yes, I'm still getting the bug.

--- 
Distribution: Arch Linux
GNOME Shell 44.1
Display server: Wayland
PaperWM branch/tag: fixes-gnome-44-support
PaperWM commit: 477e87a4f04a01ad05245852c0cf645fd19bec0c
Enabled extensions:
- paperwm@hedning:matrix.org

**Comment by letavocado** [comment]
I added custom keybinding `Super+Backslash`(but forgot to remove the default keybinding `super+shift+above_tab`) still the same bug. I restarted system and the bug is gone. I opened the workspace preferences to remove default keybinding `super+shift+above_tab` and suddenly after removing the bug immediately appeared, without a logout and lock/unlock. And removed default keybinding still continues to work.

**Comment by letavocado** [comment]
`Super+Above_Tab`, which is supposed to cycle through the last workspaces, still also moves active windows, taking them with it.

**Comment by letavocado** [comment]
I don't know if this is helpful?

![image](https://github.com/paperwm/PaperWM/assets/31704887/6df4438d-562f-4226-9377-22a0067e704c)

**Comment by jtaala** [comment]
> `Super+Above_Tab`, which is supposed to cycle through the last workspaces, still also moves active windows, taking them with it.

Oh man.  I had this, but it was a conflict that showed in the keybind settings pane.  Can you send me a picture of your workspace keybinds (with none expanded), especially of the workspace bindings? - like this, here's mine:

![image](https://github.com/paperwm/PaperWM/assets/30424662/712074e8-a416-47ff-a084-3c549807c7f2)

I'll put in the other PR comments to see who else is seeing this stuff.

**Comment by jtaala** [comment]
> The `Super+PageUp` does not work at all, only the `Super+PageDown`

Yeah, this one is due to a default gnome keybind conflict, which I had to disable:

![image](https://github.com/paperwm/PaperWM/assets/30424662/d2ae0427-236a-4846-995e-b367a3cc8ba5)

Okay, I think I'm going to have to implement some specific disabling for clashes in PaperWM.

**Comment by jtaala** [comment]
Found the clash!

It's a default gnome `Navigation` keybind clash... `Switching windows of an application` - you need to disable it.

It's the bottom one:

![image](https://github.com/paperwm/PaperWM/assets/30424662/871a71b5-9c6f-430c-9045-e80bb673f03f)

**Comment by jtaala** [comment]
I'll send you a command to disable that one.  Give me a sec.

**Comment by jtaala** [comment]
I had it disabled so the bug wasn't occurring for me.

**Comment by jtaala** [comment]
Run this command from terminal to disable the gnome 44 keybind that clashes:

```
gsettings set org.gnome.desktop.wm.keybindings switch-group "[]"
```

**Comment by jtaala** [comment]
Ha, so see https://github.com/paperwm/PaperWM#keybindings

Which actually talks about the `super+above_tab`.  Question is, why didn't it override your settings properly in gnome 44 for you?

Will need to think about that one.  One thing I remember, is that for gnome 44, when I first installed PaperWM - I recall a popup asking if I wanted to allow extensions to override keybindings (or something like that)... I had to click yes.  Do you remember seeing anything like that?  The first time I saw that was in gnome 44 I think.

**Comment by jtaala** [comment]
Hey @letavocado, I can finally reproduce this now.  Thanks for the patience, I'll work on a fix for this and get you to test it when I figure it out.

Cheers.

**Comment by Lythenas** [comment]
@jtaala Should we just disable the conflicting gnome key bindings in our `set-recommended-gnome-shell-settings.sh`? Or do you want to implement some other checks?

Maybe we can also implement a general check for conflicting keybindings (maybe either on startup or displayed in the preferences window).

**Comment by jtaala** [comment]
> @jtaala Should we just disable the conflicting gnome key bindings in our `set-recommended-gnome-shell-settings.sh`? Or do you want to implement some other checks?
> 
> Maybe we can also implement a general check for conflicting keybindings (maybe either on startup or displayed in the preferences window).

Hey @Lythenas, yes, I think so.  I was digging into why this issue was occurring (I don't recall previously versions having this issue).  It may be due to changes in gnome 44 - in any case, I think we disable them in the install script and I'm also thinking about

- disable these on paperwm enable;
- maybe add a few buttons to the keybinds settings, something like `disable clashing gnome keybinds` and `restore clashing gnome keybinds` or something?

Could you also add the following other clashes I found?
```
gsettings set org.gnome.desktop.wm.keybindings switch-to-workspace-left "[]"
```
```
gsettings set org.gnome.desktop.wm.keybindings switch-to-workspace-right "[]"
```
These clash with the default PaperWM keybinds for `Switch to workspace above` and `Switch to workspace below`.

Thanks!

**Comment by letavocado** [comment]
Hey, @jtaala. Sorry, I've been very busy at work. :pray: 

> Oh man. I had this, but it was a conflict that showed in the keybind settings pane. Can you send me a picture of your workspace keybinds (with none expanded), especially of the workspace bindings? - like this, here's mine:

I'm using default keybindings. Except `Super + Backslash`, that's only for testing purposes.
![image](https://github.com/paperwm/PaperWM/assets/31704887/c18f6cc3-e559-4f0c-948d-a8d662f76c76)



> Run this command from terminal to disable the gnome 44 keybind that clashes:
> 
> ```
> gsettings set org.gnome.desktop.wm.keybindings switch-group "[]"
> ```

Thanks, @jtaala! Cool, that's fixed the bug. But now <kbd>Super</kbd><kbd>Shift</kbd><kbd>Above_Tab</kbd>, moving active windows.


 

> Which actually talks about the `super+above_tab`. Question is, why didn't it override your settings properly in gnome 44 for you?
> 
> Will need to think about that one. One thing I remember, is that for gnome 44, when I first installed PaperWM - I recall a popup asking if I wanted to allow extensions to override keybindings (or something like that)... I had to click yes. Do you remember seeing anything like that? The first time I saw that was in gnome 44 I think.

Hmm... Perhaps I saw, but was not quite attentive, something clearly popped up. But it didn't show up again.

> Hey @letavocado, I can finally reproduce this now. Thanks for the patience, I'll work on a fix for this and get you to test it when I figure it out.

Great, I'm ready. :heart: 

Thanks!

**Comment by letavocado** [comment]
> Thanks, @jtaala! Cool, that's fixed the bug. But now <kbd>Super</kbd><kbd>Shift</kbd><kbd>Above_Tab</kbd>, moving active windows.

If I remove custom keybinding <kbd>Super</kbd><kbd>Backslash</kbd> and reset to default keybindings. <kbd>Super</kbd><kbd>Shift</kbd><kbd>Above_Tab</kbd> works as expected, without moving active windows. But, if I add custom keybinding again and remove default one(**required to reproduce this bug**), it starts moving windows again.

Custom keybinding <kbd>Super</kbd><kbd>Backslash</kbd> doesn't move windows.
Default keybinding <kbd>Super</kbd><kbd>Shift</kbd><kbd>Above_Tab</kbd> moves.

**Comment by jtaala** [comment]
> Thanks, @jtaala! Cool, that's fixed the bug. But now SuperShiftAbove_Tab, moving active windows.

Oh lol.  This will fix that one:

```
gsettings set org.gnome.desktop.wm.keybindings switch-group-backward "[]"
```

**Comment by jtaala** [comment]
So far, it's these keybinds that have clashes and we can add to the recommended settings script.

1. `gsettings set org.gnome.desktop.wm.keybindings switch-group "[]"`
2. `gsettings set org.gnome.desktop.wm.keybindings switch-group-backward "[]"`
3. `gsettings set org.gnome.desktop.wm.keybindings switch-to-workspace-left "[]"`
4. `gsettings set org.gnome.desktop.wm.keybindings switch-to-workspace-right "[]"`

**Comment by jtaala** [comment]
Hey @letavocado, can you pull the latest #515 (e.g. commit 3d0f25e2d12d2fb18be090e33e52b3a0a40b030d)?  PaperWM now disables the above clashes when it's enabled.  These should fix this issue.

@Lythenas - see  https://github.com/paperwm/PaperWM/pull/515/commits/3d0f25e2d12d2fb18be090e33e52b3a0a40b030d this now disables the clashes that were in the recommended settings script... it does this when PaperWM is enabled.

I haven't touched the script, but these changes make the script settings superfluous (so I suppose they can be removed from the script at some point).

**Comment by Lythenas** [comment]
> @Lythenas - see [3d0f25e](https://github.com/paperwm/PaperWM/commit/3d0f25e2d12d2fb18be090e33e52b3a0a40b030d) this now disables the clashes that were in the recommended settings script... it does this when PaperWM is enabled.

I don't think we should do it like this. Because if the user has custom keybindings for the "known clashes" they would be disabled even if they don't actually clash. Instead  I would just add them to the set-recommended-gnome-shell-settings script. And if you want to disable clashes on PaperWM enable, I would check for actual clashes. (Also I guess we should restore the keybindings when we disable PaperWM.)

I think I also wouldn't change the keybindings automatically because the gsettings are stored on disk, right? So the settings are still changed even if the user uninstalls or disables PaperWM. I think adding a list of conflicting keybindings to the PaperWM settings would be better (but not sure how hard to implement).

I will add the conflicting keybindings to the set-recommended-gnome-shell-settings script and make a PR. And I think everything else can be improved after we have gnome 44 running.

Edit: PR #525 Might also be worth merging this in the gnome-43 branch.

**Comment by letavocado** [comment]
Thanks @jtaala! Pulled! Awesome!

**Comment by jtaala** [comment]
> > @Lythenas - see [3d0f25e](https://github.com/paperwm/PaperWM/commit/3d0f25e2d12d2fb18be090e33e52b3a0a40b030d) this now disables the clashes that were in the recommended settings script... it does this when PaperWM is enabled.
> 
> I don't think we should do it like this. Because if the user has custom keybindings for the "known clashes" they would be disabled even if they don't actually clash. Instead I would just add them to the set-recommended-gnome-shell-settings script. And if you want to disable clashes on PaperWM enable, I would check for actual clashes. (Also I guess we should restore the keybindings when we disable PaperWM.)
> 
> I think I also wouldn't change the keybindings automatically because the gsettings are stored on disk, right? So the settings are still changed even if the user uninstalls or disables PaperWM. I think adding a list of conflicting keybindings to the PaperWM settings would be better (but not sure how hard to implement).
> 
> I will add the conflicting keybindings to the set-recommended-gnome-shell-settings script and make a PR. And I think everything else can be improved after we have gnome 44 running.
> 
> Edit: PR #525 Might also be worth merging this in the gnome-43 branch.

Thanks @Lythenas!

Good points.  However, I believe I've actually found the root cause issue here (see commit  1ac33b4620c1cecc9578eab50b92fbe2ee109673).

PaperWM has always had keybind conflict detection and overriding.  I just couldn't figure out why it wasn't working properly.  It's supposed to check for conflicts, and override them while PaperWM is enabled/active.  It does this by checking the schema keycombos of the system against PaperWM's keybinds, and if a conflict is found it will override (non-persistently) while PaperWM is on.

Long story short - there was a function signature mismatch which basically caused the key combo mask to not be interrogated properly, basically to be ignored, meaning that conflicts weren't being identified properly (see [here]( https://github.com/paperwm/PaperWM/pull/515/commits/1ac33b4620c1cecc9578eab50b92fbe2ee109673#diff-2fad2176cebef6b0a034da4cc5506fefd8a02a9654398e5a72b36fc5b9991323L245)!).

I believe this has been broken for a while - and since gnome 44's keybinds, the issue showed itself better.

Good new is, that I now don't believe we need the settings overrides in the recommended scripts.

@letavocado, once pulled the latest, you should be able not disable the system keybinds that were causing this.

**Comment by jtaala** [comment]
I've since removed the previous "fix" (e.g. `disableKnownClashes`).
